Prevent UID translator from changing perms outside volume

novas0x2a opened the pull request:
When a volume had a symlink, the symlink would be resolved outside the
volume during UID/GID translation. The os.Walk used during translation
would call Lstat on each object, and this is what is used during
translation. This means that the Mode() would be for the symlink itself
(which is 777). However, the chmod call would act through the symlink,
changing the permissions on the underlying object.

This meant that any symlink in the volume would result in a chmod 0777
call directed at a file on the host system itself, allowing a hostile
volume to change permissions on important system files (/usr/bin,
/etc/passwd, etc) by creating those files.

The most correct fix here would be a lchmod call, but that doesn't exist
because POSIX assumes symlinks don't have independent permissions (and
so are always 0777). This means this fix should be correct on platforms
for which this assumption holds; however, this assumption does not hold
on OSX. I am unaware of any high-level go API that solves that problem,
which means a more nuanced fix here would require a direct syscall on
OSX (see https://github.com/golang/go/issues/20130 for some discussion).

concourse/concourse#1736

pivotal-issuemaster commented:
@novas0x2a Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/baggageclaim&pullRequestId=10)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=concourse/baggageclaim&pullRequestId=10) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@novas0x2a Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=concourse/baggageclaim&pullRequestId=10)!

vito commented on integration/baggageclaim/privileges_test.go:
```diff
@@ -57,16 +83,19 @@ var _ = Describe("Privileges", func() {
 		runner.Stop()
 		runner.Cleanup()
+
+		checkMode(linkSentinel, sentinelMode)
```
It seems kind of non-obvious that this should be something done for every test in the file - could it be isolated to a single, new `It` instead?

vito reviewed (changes requested):
Thanks for submitting! I'd like to see an explicit test for this though rather than modifying the existing context.

novas0x2a committed:
Prevent UID translator from changing perms outside volume

When a volume had a symlink, the symlink would be resolved outside the
volume during UID/GID translation. The os.Walk used during translation
would call Lstat on each object, and this is what is used during
translation. This means that the Mode() would be for the symlink itself
(which is 777). However, the chmod call would act through the symlink,
changing the permissions on the underlying object.

This meant that any symlink in the volume would result in a chmod 0777
call directed at a file on the host system itself, allowing a hostile
volume to change permissions on important system files (/usr/bin,
/etc/passwd, etc) by creating those files.

The most correct fix here would be a lchmod call, but that doesn't exist
because POSIX assumes symlinks don't have independent permissions (and
so are always 0777). This means this fix should be correct on platforms
for which this assumption holds; however, this assumption does not hold
on OSX. I am unaware of any high-level go API that solves that problem,
which means a more nuanced fix here would require a direct syscall on
OSX (see https://github.com/golang/go/issues/20130 for some discussion).

concourse/concourse#1736

novas0x2a commented:
We talked about it on slack, but just making sure this is good to go :)

vito commented:
Looks good, thanks for the reminder!
